processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java

```diff
@@ -137,85 +186,130 @@ public void write(T objectToWrite) throws IOException
     );
     Preconditions.checkState(
-        numBytesWritten < fileSizeLimit, "Wrote[%s] bytes to base file %s, which is too many.",
-        numBytesWritten,
-        filenameBase
+        numBytesWritten < fileSizeLimit, "Wrote[%s] bytes, which is too many.",
+        numBytesWritten
     );
 
-    try (OutputStream metaOut = ioPeon.makeOutputStream(makeFilename("meta"))) {
-      metaOut.write(GenericIndexed.VERSION_ONE);
-      metaOut.write(objectsSorted ? GenericIndexed.REVERSE_LOOKUP_ALLOWED : GenericIndexed.REVERSE_LOOKUP_DISALLOWED);
-      metaOut.write(Ints.toByteArray(Ints.checkedCast(numBytesWritten + 4)));
-      metaOut.write(Ints.toByteArray(Ints.checkedCast(numWritten)));
-    }
+    ByteBuffer meta = ByteBuffer.allocate(metaSize());
+    meta.put(GenericIndexed.VERSION_ONE);
+    meta.put(objectsSorted ? GenericIndexed.REVERSE_LOOKUP_ALLOWED : GenericIndexed.REVERSE_LOOKUP_DISALLOWED);
+    meta.putInt(Ints.checkedCast(numBytesWritten + Integer.BYTES));
+    meta.putInt(numWritten);
+    meta.flip();
+
+    Channels.writeFully(channel, meta);
+    headerOut.writeTo(channel);
+    valuesOut.writeTo(channel);
   }
 
-  private void closeMultiFiles() throws IOException
+  private void closeMultiFiles(WritableByteChannel channel, FileSmoosher smoosher) throws IOException
   {
-    headerOutLong.close();
     Preconditions.checkState(
-        headerOutLong.getCount() == (numWritten * Longs.BYTES),
-        "numWritten[%s] number of rows should have [%s] bytes written to headerOutLong, had[%s]",
+        headerOutLong.size() == numWritten,
+        "numWritten[%s] number of rows doesn't match headerOutLong's size[%s]",
         numWritten,
-        numWritten * Longs.BYTES,
-        headerOutLong.getCount()
+        headerOutLong.size()
     );
     Preconditions.checkState(
-        headerOutLong.getCount() < (Integer.MAX_VALUE & ~PAGE_SIZE),
-        "Wrote[%s] bytes in header file of base file %s, which is too many.",
-        headerOutLong.getCount(),
-        filenameBase
+        (((long) headerOutLong.size()) * Long.BYTES) < (Integer.MAX_VALUE & ~PAGE_SIZE),
+        "Wrote[%s] bytes in header, which is too many.",
+        (((long) headerOutLong.size()) * Long.BYTES)
     );
+
+    if (smoosher == null) {
+      throw new IAE("version 2 GenericIndexedWriter requires FileSmoosher.");
+    }
+
+    int bagSizePower = bagSizePower();
+    ByteBuffer meta = ByteBuffer.allocate(metaSize());
+    meta.put(GenericIndexed.VERSION_TWO);
+    meta.put(objectsSorted ? GenericIndexed.REVERSE_LOOKUP_ALLOWED : GenericIndexed.REVERSE_LOOKUP_DISALLOWED);
+    meta.putInt(bagSizePower);
+    meta.putInt(numWritten);
+    meta.putInt(fileNameByteArray.length);
+    meta.put(fileNameByteArray);
+    meta.flip();
+    Channels.writeFully(channel, meta);
+
+    long previousValuePosition = 0;
+    int bagSize = 1 << bagSizePower;
+
+    int numberOfFilesRequired = GenericIndexed.getNumberOfFilesRequired(bagSize, numWritten);
```

Comment:
Please add some comments what's going on here.

Reply:
Not sure what I should comment here